feat(balance): Adjusting mass, thrust, and turning across all ships

Zitchas opened the pull request:
**Balance:**

## Summary
- All engines, thrusters, and steering increased.
- Most ships increased in mass. Indigenous, entirely self-contained unique ships (like the Shooting Star), and special ships (like timers) have not been affected.
- Most fighters, drones, interceptors, and similarly sized ships have not had any increase in mass.
- Most ships larger than an interceptor have had their mass increased.


Here's the notes provided by Arkhne on the values and scope of the changes:

<details>
<summary>Arkhne's notes</summary>

Drone 		- 1
Fighter		- 1
Interceptor	- 1
Light Warship	- 2
Medium Warship	- 4
Heavy Warship	- 8


Human Ships

Light	(120) 130     	-	130 (250)	
Medium	(130) 170	-	310 (880)
Heavy	(500) 600  	-	880 (910)

120-150-500-1000


Wanderer

Light	(###) 110	-	130 (240)
Medium	(###) 260	-	320 (###)
Heavy	(###) 540	-	770 (###)

110-250-500-1000


Hai

Light	(###) ###	-	220 (###)
Medium	(168) 280	-	320 (###)
Heavy	(###) 650	-	925 (###)

110-200-500-1000


Remnant

Light	(###) ###	-	270 (###)
Medium	(###) 350	-	350 (###)
Heavy	(###) ###	-	610 (###)
Dromedary - Special Case, x6.

100-300-500-2150


Coalition [Heliarch]

Light	(140) 180	-	270 (280) [270]
Medium	(340) 350	-	380 (###) [520]
Heavy	(###) 570	-	640 (670) [780]

110-300-530-1000


Bunrodea

Light	(###) ###	-	250 (370)
Medium	(###) ###	-	580 (###)
Heavy	(###) 650	-	### (###)

130-380-600-2000


Gegno

Light	(###) 230	-	275 (###)
Medium	(238) ###	-	740 (###)
Heavy	(###) 1032	-	1090 (###)
Augen - Special Case, x6

150-370-1000-2000


Korath

Light	(###) 250	-	370 (420)
Medium	(540) 580	-	690 (780)
Heavy	(970) 979	-	1745 (1755)
Ikatila'ej - Special Case, x6
Korsmanath - Special Case, x5

150-500-960-2000

</details>

## Reasoning
Combat balance in Endless Sky has been a constant problem: It's been more or less considered to be "good enough" for years now, while simultaneously being a constant source of complaints. This PR aims to address some of these complaints, and provides the foundation for a few additional elements that should round them off and provide for a new and interesting combat environment

Some of the complaints that this PR hopes to solve:
- There's very little spread in mobility of ships.
- Larger ships can often go faster (and even accelerate faster) than small ships.
- No point in flying fighters and interceptors when it's possible to get a heavy warship to be just as manoeuvrable.
- It's boring when everything pretty much goes the same 300-600 speed.

## Relevant links:
https://github.com/endless-sky/endless-sky/issues/4034
https://github.com/endless-sky/endless-sky/pull/5772

(these are but two of many)

## Save File
This affects all parts of the game, so any save file at any point in any of the campaigns will be useful. The main caveat is that one absolutely **must** purchase a new ship and start a new mission; as any existing ships (whether player owned or NPCs) that are saved in the pilot file will be using previous mass values, and thus will have wildly inaccurate performance. 

### You must purchase new ships and start fresh missions to test this PR as it changes ship mass values, and these will only come into effect with NEW ships, either purchased or created by jobs/missions accepted after starting with this PR.


## To-do:
- [ ] Verify that none of the ships are too fast; and that none are too slow. We want a nice large spread; but we neither want fast ships to be uncontrollably fast; nor do we want slow ships to be painfully lethargically slow. Note: It is perfectly acceptable for some, even many, ships to be too slow or too fast for some people. Or even for most people. For example, fighters and the heaviest of heavy freighters. It's generally assumed that most people aren't going to fly fighters; and given the combat nature of the game, most people aren't going to want to fly almost helpless freighters either. Having some ships that exceed useful speeds are useful for adding depth and detail to the game. But they shouldn't be *too* extreme...
- [ ] Tweak values as needed to polish it.
- [ ] Extensive play test by people willing to set their preconceptions and favorites aside and just *play* it.

Zitchas commented:
This branch/PR is never going to be PR'd by itself to mainstream, so I "updated" it to experimental. Should be much more testable now.

Zitchas commented:
One thing worth noting: the drag value has not been adjusted, so many ships are quite a bit faster than they used to be. While this is something that needs to be looked at, I'm also not too concerned about it, as generally speaking a ship's acceleration is much more important than its top speed. Also, realistically speaking, ships have infinite maximum speed, just different rates of change. (until they get into fractions of the speed of light, anyway)

Anyway, higher speeds can be useful for allowing larger systems and promoting the spread of ships, while reducing its impact on general travel. (all ships can get up to a reasonably fast "cruise" speed for crossing large systems, minimizing negative impact; meanwhile fast ships are advantaged because they can get up to their cruise much faster, and can change direction more quickly.

Zitchas commented:
Note: This most recent update included https://github.com/endless-sky/endless-sky/pull/8842 . In resolving it's conflicts, I used it as the basis for doubling all its values (the same pattern that Arkhne used initially) in order to keep the new values in line with the mass changes. Arkhne commented that the commentary on the flattening PR made it sound similar to what they originally did back in the first Revamp. (We haven't compared numbers, though, so we don't know how close it actually is)

bene-dictator commented:
I'm too busy to test this and give good feedback at the moment but I think if this was posted to the community discord server we would see more feedback... additionally with executables to allow players that aren't familiar with GitHub / compiling to playtest.

It's a shame we don't have a `#ES-Delta` channel under community projects in the server, too. I'd be down for one.

Zitchas commented:
If you would like an ES-Delta channel on discord, then I would recommend you suggest that to one of the moderators on Discord. If you want to share a link to this, you are welcome to, but I would prefer that discussion on its details happens here and not discord. Discord has already had many years of debate on this topic, much of which informed the work that has gone into this; along with the discussion that accompanied the first Revamp. If you know of people who would like to playtest something like this, then that's great! playtesting is really what is need the most right now.

Personally, I would like to get Delta a bit more polished before it is subject to the full treatment: While this piece is a major element, it's still just one of several parts of a combat revamp, and none of them individually solve the problem, and some haven't been implemented yet. It's really a case where many interlocking components all need to change together in order to really get the combat environment that we want. Which is to say, a varied environment where many combat styles are possible, where small ships are actually useful for the player to fly in larger engagements, and "biggest warship you have access to" is not the automatic mandatory ship you must be flying.

At this point, and particularly for informing refinements to this PR, personal experience play testing this specific product is... perhaps not actually *required*, but definitely something that is strongly encouraged to ensure people really have a knowledge of what they are commenting on. Of particular note is that this PR does not exist in a vacuum. It is built on a range of previously added elements such as eliminating the majority of drag from flight characteristics, the addition of lateral thrust, and some other factors that make this much closer to being an actual *space* flight game and less of a "high speed ships on water" navigation. As such, people cannot just use what they know of regular vanilla ES, just with different masses and thrusts. That'll be close, but not the whole picture, and the details do make for some subtle but overall significant differences.

---

That being said, the full ES zip file produced by the checks is available, and is the safest and most reliable way to download it, as it is produced by github's routines. Executables uploaded by me or anyone else should be treated with extreme caution and verified using a virus scanner before being allowed to run. 

To download the products of the checks:
1. look in the checks box at the bottom of the PR. (It should say "All checks have passed", with "X successful and Y skipped checks". Skipped is OK, that means that check wasn't relevant. If it says failed, though, there's stuff to be resolved. Or one of github's checks is messing up.)
2. Click on the "Details" link next to the OS of choice (ex. "CD / Windows (pull_request)")
3. Click on the "Summary" button that should be on the left side near the top.
4. Somewhere on the page, usually towards the bottom, there should be a section titled "Artifacts" with the subtext "produced during runtime."
5. There should be at least three artifacts here, one each for Linux, MacOS, and Windows. Download the zip that will be most useful for you. 
6. Unzip that zip somewhere, then unzip the zip that it inside it. This should give you a fully functional portable Endless Sky installation with the contents of the PR you downloaded it from, including the appropriate exe.

Note 1: The Delta experimental development version of Endless Sky has lateral thrust, so you will need to go into your preferences and assign the left and right thrust to appropriate keys before you will be able to use them.

Note 2: This PR changes actual ship data. Due to the fact that endless sky saves ship data (but not outfit data) to the player file, no ship that is already in the player file will reflect any of the changes of this PR to the ship. This includes both player and NPC ships. So while you can use old saves, you *must* use new ships and missions that were accepted *after* starting to use this PR in order to test it.

Note 3: In code PRs, the same process, but clicking on a different "details" link in step 2 will take you to a similar location, but that only downloads the executable, not the whole installation. This is one of the "skipped checks" for this particular PR as this one doesn't make any code changes.

Zitchas commented:
Notes of things to look into:

Misc thoughts:
- [ ] For categories that are generally internally consistent in terms of mass, which is to say most categories other than "Transport" and "Utility", we should take a look at drag values and consider adjust to level out their theoretical max speed a bit. Just thinking that, for instance, all the light freighters should probably have comparable max speeds
- [ ] In Sea of Thieves, they balanced the three ships in an interesting way: The smallest ship can turn faster and sail against the wind faster than the big ship, but the big ship can sail *with* the wind faster than the small ship. Kind of analogous to saying that the big ships have higher max speed, but the smaller ships have better acceleration and turning. In an out and out race, the big ships will win; but if the goal is to catch the small ship, the small ship can escape by zigzagging and forcing the big ship to continuously have to re-accelerate again in the opposite direction, thus ensuring it spends less time at its max speed, and nullifying its advantage. This works especially well if acceleration values are such that most ships (especially big ships) will need to accelerate for quite a few seconds to reach their max speed. As in, 10 seconds or more.

---
Ship comparison. Vanilla to this PR comparison, and thoughts deriving therefrom:
- [x] Lance seems too expensive (partially solved by reducing supercapacitor price, potentially also outfitting change currently PR'd
- [ ] Argosy. Lore has lots of extra engine and customization space, and Image has giant quartet of engines, contrast to Hauler which does not, yet the two have the same engine space and by virtue of argosy being heavier, hauler has more aceleration potential. (argosy has strong speed advantage) Hauler even has more hull than the Argosy, with the later having more shields. I'd expect hull to be heavier than shields... Argosy also has less 10 less cargo but 20 more outfit space. All in all, I'm not sure the differences merits the 240 ton heavier mass of the Argosy.
- [ ] Light Freighters in general: Star barge 50 cargo, Clipper 70, Argosy 120, Hauler 130, Freighter 150. Also the sprite sizes. I'd say there's a strong case to be made that the Argosy, Freighter, and Hauler should be moved into a "Medium Freighter" category.
- [ ] Heavy Freighters: All in all, it's glorious demonstration of slow and steady. One thing that comes to mind, though, is "where is that mass coming from?" These ships really don't have a lot of shields or hull, nor a particularly large amount of bunks or outfit space or anything either. Makes me think that maybe the hull mass for these ships should be decreased by a fairly large amount - but then their cargo space *increased* by about the same amount. These are the workhorses that keep the economy running, right? They should be delivering a lot more cargo. Thought: Triple the amount of cargo space, and remove that increase from their mass. A fully empty train (or tractor trailer) is actually kind of impressive in performance, and a ship that is dedicated to the brim for cargo should feel *massively* different between when fully loaded and fully empty. (for a point of comparison: An empty 40' sea can weighs  3,750kg, and its maximum full weight can reach 26,300kg).
- [ ] The Hauler series in particular is described as basically a bunch of cargo containers with a cockpit stuck on one end and engines on the other. These ships should be very heavily cargo dependant for their mass.
- [ ] Berserker's turning rate is... kind of crazy. 304-315. On the one hand, I can't handle turning that fast. But on the other, I'm not sure I could handle an F-16, either. Having some ships beyond the ability to handle of some players isn't a bad thing. The fact that we can use ctr to halve the rate of turn helps a lot.

And Interceptors is where I got to for now.

Saturnalia9 commented:
> One thing worth noting: the drag value has not been adjusted, so many ships are quite a bit faster than they used to be. While this is something that needs to be looked at, I'm also not too concerned about it, as generally speaking a ship's acceleration is much more important than its top speed. Also, realistically speaking, ships have infinite maximum speed, just different rates of change. (until they get into fractions of the speed of light, anyway)
> 
> Anyway, higher speeds can be useful for allowing larger systems and promoting the spread of ships, while reducing its impact on general travel. (all ships can get up to a reasonably fast "cruise" speed for crossing large systems, minimizing negative impact; meanwhile fast ships are advantaged because they can get up to their cruise much faster, and can change direction more quickly. 

Ships will need considerably higher turning (or have high built-in reverse thrust to engines or ships) to accomodate high max speeds. It's painful and rage-inducing to wait 10 seconds just to wait for a ship to turn around, and then another 5 or 6 seconds to decelerate. I've not tested this experiment yet, but that's one of the main reasons why high max speed and low acceleration aren't too compatible.

Also note that I don't use vanilla stats for my playthroughs, so my slow is slower and fast is genuinely faster than vanilla.

Saturnalia9 commented:
> - [ ] Berserker's turning rate is... kind of crazy. 304-315. On the one hand, I can't handle turning that fast. But on the other, I'm not sure I could handle an F-16, either. Having some ships beyond the ability to handle of some players isn't a bad thing. The fact that we can use ctr to halve the rate of turn helps a lot.

When we (Doom and I) were talking about the different types of speeds, we generally agreed that 200-250 turning is more ideal than 300, since the player will be steering these ships manually. Sure, there is auto-pilot and auto-target, but we still have to factor in people who don't use auto-target in battles. 

Humans don't often (if at all) have a response time within 0.2 seconds. Think the record was 0.5? So you'd have an angle of ~150 degrees per key tap. That's a fair amount to try to angle your guns with. So we wanted to skirt around that issue a bit, by instead putting ships in a specific range of agreeable stats.

Fighters and drones _usually_ aren't piloted by the player (unless you really don't like yourself), so we allowed absurd uncontrollable speeds for them wherever needed.

Zitchas commented:
A couple thoughts here:
1) Given that we implemented universal lateral thrust in view of the fact that station keeping is a universal spaceship capability (as in, the ability to make minor course corrections and changes to position to remain in orbit or otherwise in a fixed position, mitigating minor forces that might be acting on the ship); I suspect that universal reverse thrust is likely too. Probably a pretty minor amount (5-10% of forward thrust as reverse thrust on all thrusters).

2) There's been a fair amount of discussion about expanding system sizes. VitalChip has an extreme example of this where Sol is *vastly* bigger than our current one; resulting in quite a bit of in-system traffic flying around. Sometimes one can even note traffic patterns. At this point, I'd say there feels like a fairly good degree of agreement that systems should be bigger, we just don't know how *much* bigger. 

The reason this is relevant is that expanding systems changes the combat dynamic from "each system is one combat location" to "each system is a combat region, with ships and fleets moving around within it. The FW could be getting massacred on this side of the system, while completely dominating in a different area. Bigger sizes, and more places to land per system mean that having a max speed that takes many seconds to reach is actually useful. It also means that most ship's "combat speed" is going to be more along the lines of 2-3x their acceleration, with their actual max speed being something that is generally only reached when cruising between two distant points. And yes, this does mean that ships will need to do Expanse style manouvers where they burn across the system, then spend the last several seconds approaching combat pointing backwards while they decelerate so that they don't just zip past the combat entirely. Which is a pretty big shift compared to the current combat paradigm of "point towards target, hold down accelerate and fire. Hold until dead."

Anyway. Bigger systems, as well as departure distances, means that having a higher max speed even on ships with low levels of acceleration will not only be useful, but is also fairly essential in keeping big ships playable. No-one wants to wait several minutes to reach the hyperjump limit. But having to wait 5-10 seconds to get up to full speed to cruise out there while watching hawks and other light and agile ships get up to cruise in 2-3 helps emphasize the differences between sizes of ships, even if the two have not-so-disimilar actual cruise speeds. In space, velocity is precious, and Delta V even more so. This is something that Endless Sky has always ignored, to our detriment. 

3) But yes, I agree that >250 turning speeds are probably too fast. Once they reach the realm of *no* human being able to use it (as opposed to just some players that are really good with it), then it should be adjusted. Probably by swapping out its outfitting, but that's hard to say. That list above is, as noted by being a checkbox list, a sort of overview "things I need to look at more closely".

4) I'll just explicitly call out the ctr key modifier here, as I feel it isn't very well known. In Delta we implemented the ctr key (while in flight) as a "half input" modifier that can be applied to any of the movement commands. If you turn at 300, then while holding the ctr key, you will now turn at 150. If you accelerate at 800, with the ctr key you accelerate at 400. Applies to turning, thrusting, reverse thrusting, and lateral thrust. It takes some getting used to, but it's very helpful in allowing the player to precisely control high performance ships.

Quantumshark commented:
As a more intuitive - and probably also more useful - alternative to holding ctrl for fine maneuvering control, how about adding an 0.5x speed setting, behaving similarly to the 3x speed setting that already exists? That would achieve the same goal (easy control of ships with 200-400 turning), but would also make it far easier to avoid projectiles, work around the rotation of turrets and other ships, etc. My concern about the ctrl for half mobility approach is that it puts the player at a disadvantage against their opposed counterparts, which don't need to make use of it. What's the point of that extra speed if all it does is means you have to hold down a key that disables it in order to make the ship controllable? 0.5x game speed, on the other hand, would make the player's ship controllable, but would also slow down everything else that was happening by the same amount, which makes that extra speed far more useful.

That said, I have noticed that with very high acceleration stats, the amount of turning that is desirable, and/or comfortable, is noticeably higher, so it may be that 200+ turn rates would be fine as-is once people have time to get used to them. I do think it's an issue that the extremely mobile ships that favour an experienced pilot with lightning-fast reactions are the ones forced on a new player, while the less demanding ships are the ones that are locked to later in the game, but I don't have any solutions for that particular problem.

I'll also link https://github.com/endless-sky/endless-sky/discussions/8957 here, since it's my concept for a similar change - with the key difference between that and this being making changes such that acceleration and turn are decoupled, since the range of variety I'd like to see in acceleration is much wider than I think would be ideal for turning. With acceleration changes, it'd be reasonable and enjoyable if heavy warships had accelerations of generally 50-150, while fighters had accelerations of 2000 and up. For acceleration, that much range is desirable. For turn, that much range is too wide - ideally, turn rates below about 50 should be the exception, *not* the rule - something reserved for superheavies or very unusual setups, whereas turn rates past 500 are completely outside the realm of being human-controllable - so with that wide a range of turn rates, at least one end is going to be completely unplayable, quite possibly both ends depending on where you centre the range. Based on this and on my proposal in https://github.com/endless-sky/endless-sky/discussions/8957, I'd suggest steepening the size scaling of steering engines, but to a lesser extent than you've steepened the scaling of masses.

One thing here that I do strongly disagree with is that you're pushing for significantly greater homogeneity within categories. I don't think it's appropriate to apply the same changes to the Vanguard as to the Carrier, for example - those are very clearly two extremely different ships, and the fact that they happen to both be categorised as `heavy warship` should not be seen as an invitation to attempt to iron out that variety.
I also don't agree with your suggestion of tweaking drag to even out max speed within categories - in the specific case of freighters, there are certainly ships that have had their drag artificially increased because "freighters don't need to be fast" - specifically the haulers, and most heavy freighters - and it would be an improvement to undo those increases, but I don't think it'd be an improvement if all light freighters had the same top speed across the board.

Zitchas commented:
Those are some good insights and comments; particularly coming with some concrete examples of numbers. 

I will just point out that this is the starting point, and there's a reason it hasn't been merged yet. And that reason is beacause we really need to get it polished up to feel good. And, quite possibly, we need to make some changes to the way systems are sized right now so that we can settle on max speeds that are reasonable.

In terms of my comments about max speeds... I'm really aiming at "uniform top speeds" being relevant mainly for the long burns for cross-system travel. Ships absolutely should have signifiantly different feels to them whenever they are maneuvering. 

Also, feel free to suggest changes. Make the case for improved values, and I'll discuss and/or merge accordingly.

Zitchas commented:
> whereas turn rates past 500 are completely outside the realm of being human-controllable

Agreed. To expand on my previous point: I'm good with there being ships whose stock turning (or acceleration) is beyond my capabilities to effectively use; and by extension beyond many player's ability to use. But we shouldn't have any "regular gameplay" ships whose manouverability exceeds *all* human player's ability to use.

The caveat about "regular gameplay ships" being primarily aimed at fighters, indigenous lifeforms, timers, and other such ships that are either not actually intended for players to use; or flat out impossible for them to use in-game without using a plug-in or data-editing. If someone really wants to use a fighter built for the elite twitch pilots recruited from the human population... No reason to block them. But said ships also shouldn't be slowed down just so that us regular non-elites can fly them temporarily.

Saturnalia9 commented:
We were talking in Discord about ratios between max speed, accel, and turning, as well as stopping speed (whether turning 180° or reverse thrusters), and generally the acceleration to turning ratio seems to be fairly stable, with turning and accel being the same or accel being around 4 times more than turning. 

Having a very high max speed with low accel is naturally impossible for the AI to currently control, so... it's possible, but not currently ideal. And I personally enjoy being able to stop quickly, which is why I enjoy having low max speed and decent turning.

Zitchas commented:
Is it actually impossible, though? Or just makes it visible just how bad our current "point and max thrust" AI actually is?

Saturnalia9 commented:
> Is it actually impossible, though? Or just makes it visible just how bad our current "point and max thrust" AI actually is?

Shows how bad the AI currently is at handling multiple things. xD

Zitchas commented:
lol. 

Sorry I haven't had the opportunity to make any progress with this. Currently in flux with limited-to-no ability to actually *play* for the past several weeks, and unlikely to have much over the next month or so, either. Things should improve in mid-late October, though. And a big project like this... It needs actual play experience, not just poring over numbers.

MasterOfGrey commented:
I will just say for the record that I still think the best approach is to have an algorithmic baseline (as proposed here: https://github.com/endless-sky/endless-sky/pull/6481) for ship masses, rather than any arbitrary scaling on already arbitrary values, and then rebalance everything else around that.

Zitchas commented:
Your algorythm based adjustment of mass and stats is an interesting system, but I'm not really sure the outfit space changes in particular are practical. I'm not really sure what the solution is there, and I'll admit it is realisticish. But I'm rather at a loss of what to do with it. I kind of feel like it really faces two solutions:
a. all the existing outfits are scaled up so that the really big capital ship outfits really are giant capital ship sizes (like having heavy laser turrets or whatever we deem to be heavy capital ship weapons being 200+ outfit space, and large atomic engines likewise being 200-400 outfit space...
b. All the existing stuff (even the biggest generators and thrusters) become fodder for light warships; and we have to add an entire suitable-for-capital-ships suite of outfits that can only be fitted on ships that have 1000+ outfit space.

**Cargo:**
That being said, I find the cargo changes are overall great. The bulk freighter having 5670 cargo (and likewise the upscaling of cargo across the whole freighter, transport, and utility spectrum I think is likewise great. The caveat, of course, is that I rather think that the warship cargos should have been left as-is. It's already dubious enough as it is having as much cargo as they do, let alone scaling it up. A cruiser having 60 cargo isn't great in our current ecosystem, but if heavy dedicated freighters are scaled up from the multi-hundreds to now being in the thousands, then 60 suddenly shrinks in meaning and becomes just a small supply bay, not something a significant portion of a freighter's worth. 

**Bunks**
I'm not really sure about bunks. A lot of the time it appears they haven't been changed; but sometimes they've changed a lot. I don't really have an opinion there. 

All in all it's just... I don't know. It's an interesting idea, but the change in scales is hard to wrap my head around.

The other problem is that, well, it's effectively making all our ships in-scale to each other. I find that problematic because, ultimately, I really don't want our interceptors to actually be 10% of the size of a carrier. I'd much rather imagine that it's just upscaled and are actually much smaller. Likewise, that the huge ships (world ships, etc.) are actually vastly bigger than their comparative size suggests. I like having greater diversity in sizes, rather than codifying that everything to be really close together.

MasterOfGrey commented:
> Your algorythm based adjustment of mass and stats is an interesting system, but I'm not really sure the outfit space changes in particular are practical. I'm not really sure what the solution is there, and I'll admit it is realisticish. But I'm rather at a loss of what to do with it. 

It does, admittedly, present a large scope of change - though it's worth noting that those outfit space changes were determined by a factor which was calibrated to a particular intended behaviour. That factor can easily be scaled based on a different intended behaviour if desired.

> I kind of feel like it really faces two solutions: a. all the existing outfits are scaled up so that the really big capital ship outfits really are giant capital ship sizes (like having heavy laser turrets or whatever we deem to be heavy capital ship weapons being 200+ outfit space, and large atomic engines likewise being 200-400 outfit space... b. All the existing stuff (even the biggest generators and thrusters) become fodder for light warships; and we have to add an entire suitable-for-capital-ships suite of outfits that can only be fitted on ships that have 1000+ outfit space.

To be honest, it was kind of intended that you could just have *more* outfits, duplicates as necessary.
Modularisation and redundancy are important design concepts in both war and space survival.
Plus, one of the key issues I feel with trying to do things like expand missiles and counter-missile systems, or expand boarding systems, etc. are all hamstrung by how terribly tight the outfit space is on even big ships.
In contrast, it *should* be a no-brainer to choose a defence system of countermeasures against missiles (like an effective radar jammer) instead of a 5th shield generator.
Right now though, the choice is how much to downgrade your single generator on a carrier with over a hundred crew...

> **Cargo:** That being said, I find the cargo changes are overall great. The bulk freighter having 5670 cargo (and likewise the upscaling of cargo across the whole freighter, transport, and utility spectrum I think is likewise great. The caveat, of course, is that I rather think that the warship cargos should have been left as-is. It's already dubious enough as it is having as much cargo as they do, let alone scaling it up. A cruiser having 60 cargo isn't great in our current ecosystem, but if heavy dedicated freighters are scaled up from the multi-hundreds to now being in the thousands, then 60 suddenly shrinks in meaning and becomes just a small supply bay, not something a significant portion of a freighter's worth.

Warship cargo going up was more a consequence of a flat process, but I also figured that *some* larger outfits would appear as a result of these changes - and also larger ships would become horridly unwieldy for collecting loot. So having somewhat useful cargo that was still well below the actual cargo ships would allow lighter warships to collect flotsam outfits and such.

Also, part of the intent with the boarding rework plans is that storage for things like grenades (which would become consumable) would use up cargo space instead of outfit space, so equipping a heavy warship to fend off boarders would have secondary uses for that.

> **Bunks** I'm not really sure about bunks. A lot of the time it appears they haven't been changed; but sometimes they've changed a lot. I don't really have an opinion there.

Mostly the deal is that the algorithm includes a mass allowance for bunks. In this way bunks use up part of the ship's "budget" so that they're not completely arbitrary.
*Some* ships came through the algorithm with such ridiculous outcomes that modifying bunks was the most effective balancing method.
In other cases, ships intended as people movers came through the algorithm with enormous cargo holds and outfit spaces that would have made them utility cruisers instead of civilian people-movers. Subsequently "they've changed a lot" to restore their role.

> All in all it's just... I don't know. It's an interesting idea, but the change in scales is hard to wrap my head around.
> 
> The other problem is that, well, it's effectively making all our ships in-scale to each other. I find that problematic because, ultimately, I really don't want our interceptors to actually be 10% of the size of a carrier. I'd much rather imagine that it's just upscaled and are actually much smaller. Likewise, that the huge ships (world ships, etc.) are actually vastly bigger than their comparative size suggests. I like having greater diversity in sizes, rather than codifying that everything to be really close together.

Broadly speaking, you're actually asking for an even more radically extreme scaling than what I did.
Interceptors and small transports barely moved in the scheme of things. Smaller ships (most interceptors) had their sprites shrunk to fit their mass instead of their mass increased to fit their sprite (generally). There are probably some cases in there that went one way when they should've gone the other but it wasn't a refined art at the time, just a demonstration.

To pick an example; Something like the Headhunter Light Warship roughly doubled in mass, while the Carrier increased by 1,267%.
Essentially, the lightest things effectively stayed the same while the top end experienced a 12-fold increase.
(And it was already 4x larger, resulting in a final size difference of 48x the original Headhunter mass and 24x the new mass; and the Interceptors are pretty much all below the original Headhunter.)

Consider: Modern Nimitz aircraft carriers are only 20x longer than the FA-18C fighter jets that they carry.
And a lot of their mass is tied up in being a boat that can deal with the ocean.

Zitchas commented:
Ah, yes, I do see the scaling factors in there. I missed those the first time through. I sort of wonder if some of the initial values and the utter scale of the changes might have just overwhelmed me.

Too bad we don't have some sort of automated means of comparing values. I'm curious as to how the mass values in particular compare between https://github.com/endless-sky/endless-sky/pull/6481 and this PR. Particularly on a ship-to-ship basis.

One thought that occurs to me is that this PR might end up being a stepping stone towards yours. Right now I'm focused mostly on the feel of combat, the flow, and how things maneouver. The game being what it is, everything is interconnected; and there's definitely some additional factors (such as system size) that are going to feed into this too. But given time constraints, I'd rather keep outfit changes that aren't so directly tied to manouver and the flow of combat (in other words, everything other than propulsion) out for the moment. I definitely won't rule it out for the future, though. 

For instance, the fact that many ships have "free" unaccounted for stats (like bunks) is a topic that has come up fairly often in our discussions, and accounting for them so that they aren't just free bonuses for moving up to a bigger ship is important in balancing things out.

Zitchas commented:
OK, finally got a chance to spend an hour just putzing around. In this case, went and got myself a shuttle, did some flying around, and then went and found some pirates. And discovered that the AI just doesn't seem to be able to fight at all. I ended up with a firebird circling me for a couple of minutes. I wasn't even dodging, just coasting along at 57 according to my readout, and the firebird seemed either unwilling or unable to actually close with me, just orbitting around and around, constantly thrusting. I saw a falcon, a berserker, and a hawk do the same thing at other times. It honestly felt like my shuttle was almost invincible. The few times I got hit it was because other forces showed up and engaged the firebird; which changed it's movement pattern, and as a result it actually got close to me and took a few opportunistic zaps with its lasers while I went by.

The AI vs AI fights seem to go better, particularly when there's similar sorts of ships involved. When they are radically different in performance there does seem to be a fair bit of orbiting going on.

So far, I'm liking the comparison between the ships. The shuttle might be a little too zippy, but it *is* a tiny ship, so being fast was expected. Acceleration 592 might be a bit much... (feels like I'm flying a puffin...).

I just had an amusing run-in with a wasp. It attacked me, and I dodge around a bit, gave it a run for its money. Then noticed that my regen was easily keeping up with its damage. So I stopped. Now it's jousting, doing strafing runs on me. Each pass it drops me down by 15-25% of my shields, and then by the time it does a second pass, I'm back up to 90-100%, so it's just going back and forth, never making any headway...

I did get disabled by a pack of wasps, though.  So far, however, it is seeming that the shuttle is a very safe ship to fly...

Quantumshark commented:
I don't think you should have a "Medium Freighter" category if it's only going to contain two ships, but I don't see any issues with the idea other than that. I'd suggest the Water Bug, Tanto, Pelican, Dolomite, Arach Freighter and Schist for consideration.
I'd also suggest moving the Modified Argosy to "Medium Warship", if you're placing the regular Argosy in "Medium Freighter" - both for consistency, and because in all honesty that's just a much better fit for it.

Saturnalia9 commented:
Modified Argosy currently has the crew of a light warship, though.

Zitchas commented:
Yeah, I noticed the modified Argosy being a light warship, what really stood out is that it has substantially lower mass than the regular argosy. I think it is almost half the total (no cargo) mass. (700 less hull, 60 fewer cargo, and 2 fewer bunks, too.) It's sufficiently smaller than the regular Argosy that if it was a freighter it'd be a light... I must say, it's odd (and one of the few instances that I know of) where the modified variant of a ship is *lighter*...

Like I said, right now putting those two in the "Medium Freighter" category is just a means of visually separating them out. If we had a "sort shipyard by price or mass, and filter by category" option (like http://dev.endless-sky.7vn.io/), then it wouldn't be as much of an issue.

Thanks for the list. Right now I'm focused on Human space, because it's the one that has the deepest variety of ships and healthiest ecosystem. But yes, that list is probably some fairly good candidates for medium freighters.

Quantumshark commented:
I've had a chance to actually play with this a bit, so here are my initial thoughts:

The Shuttle, and other ships with 200-250 turning, were more controllable than I expected. I think the higher top speed and acceleration make the turning feel more reasonable, and the smaller sprite may also help to make it *look* like the ship isn't turning unreasonably fast.

The AI seems to struggle a lot with getting close enough to attack its targets. A lot of the time, ships just drift back and forth across each other, only dealing minor damage on each pass, and both recovering fully before they can attack again.

I started with a Shuttle, ran a few jobs, and then decided to go to Alniyat and try to capture an interceptor. After a lot of waiting and trying to lure interceptors towards militia ships, during which I could fairly easily avoid taking any damage, I was able to capture a Fury.
Re-equipped with six javelin mini pods, that Fury could easily take down other interceptors, barring the occasional one-shot from heavy rockets (which were less frequent, but also harder to predict or avoid because of how fast the ships armed with them can approach), and I even managed to outmaneuver and disable an Argosy - only to realise that doing so was pointless, because with only 15 cargo space I couldn't steal any of its larger or more expensive outfits, so I wasn't making any more profit than I would have done from attacking an interceptor. So, I still find that I don't want to just stick with an interceptor for long, but instead want to progress to light warships (which I haven't yet had time to do, so I don't have anything to say about them yet).

The very large difference in mass between the Argosy and the Modified Argosy is, in my opinion, a problem. Conserving the original difference would be fine, but now the mass difference between the two is nearly a thousand, and they seem like completely different ships. Really, I think you should apply the same mass change to both. 
I was caught out a couple of times, first by assuming the modified argosy would behave similarly to the regular one, and then once I'd realised that wasn't the case I was caught out a second time due to misidentifying a modified argosy as a regular one.

I haven't had time to start attempting story missions yet, so can't comment on how this affects those.

Zitchas commented:
Thanks for testing and the detailed feedback!

And yeah, I agree that the modified Argosy needs to be fixed. Although an amusing thought that occurred to me is that the mod argosy is basically the smallest utility class ship... lol. Actually, the regular argosy's description sounds basically like the definition of a utility ship...

Quantumshark commented:
Having acquired enough money to purchase a Headhunter, I've found that the range of ships I can defeat 1v1 hasn't really expanded much, if at all. With fewer hardpoints, I have a lower effective DPS than the Fury with six javelin mini pods did, and keeping short-range guns on target is very difficult, which leaves little room for doing so while also dodging missiles. I have tried several different loadouts (4x javelin launcher, 2x plasma cannon, 2x proton gun, 4x heavy laser), different tactical approaches, and even rebinding the keys to allow me to use lateral thrust, but I still haven't managed to disable anything LW-sized and up other than an Argosy, except in the specific case of a bounty Raven which I defeated by flying around until it had wasted its entire missile armament. On the bright side, with 50 cargo space I'm making a lot more money each time I disable an opponent, even if my actual range of targets hasn't expanded.
I'm also finding the limited outfit space on the Headhunter quite restrictive - I'm struggling to fit enough power for my engines and weapons alongside a ramscoop and the scanners I need for the story mission I've just picked up on Glaze - so I'm still not at a point where I feel I want to stick with my current ship rather than moving to a bigger one. I've set my eyes on the Splinter as the next ship to try out, and I'm definitely gathering the money for it much more quickly than I was when aiming for the Headhunter.

I've also found that combat with ships similar to my own is noticeably harder than it is without these changes - both sides are faster, which means there's less time for me to react, giving my opponents a notable advantage. An 0.5x game speed slowdown button would be extremely helpful here. Ctrl to slow my ship down just gives my opponent a bigger advantage.

Oddly, while the AI is completely incapable of engaging properly in combat, it seems not to have any issues when getting directly on top of a ship when scanning - so, the issue is definitely that the combat AI needs updating, not that any of the changes have made ships impossible for the AI to control properly at all.
The high-speed, low-control drifting of AI makes avoiding combat/damage/death very easy, especially now that I have a ship that can't be destroyed in one shot, but it makes actually engaging a target awkward to do, and AI ships seem completely incapable of engaging each other.

Looking more closely, it also seems like pirates are prioritising attacking me over attacking other merchants or the militia, to quite a significant extent. I'd expect the chance of me being targeted by any given pirate to be (1/total number of potential targets for that pirate), but in practice it seems like more than half the time all the pirates in the system are ignoring multiple merchants to target me. It's possible I'm misjudging this as a result of the aformentioned AI difficulties in targeting, but this is the impression I'm getting.

Zitchas commented:
I appreciate the testing and the detailed feedback there. Currently just getting back into the swing of things, starting with getting Delta back up to par with upstream, and likewise this PR.

I've been thinking about this, and the comments about how overall, this PR kind of makes things faster and a bit harder to control. I'm wondering if perhaps the overal thrust levels should be reduced a bit to cut all the relative speeds down somewhat. I'm thinking maybe 10% or thereabouts. I'm going to spend some time actually playing it before I do any changes though.

Zitchas commented:
I'm not sure if I've mentioned it in here; but I would very much love to have a speed control in the game. Something like the 3x button, except being 0.5x. Or just having a speed slider in the preferences. I haven't dug into the code that supports the 3x functionality, though.

The easiest solution is to just halve the speed of everything, and then let people just play the game on 3x and turn it off when they need to go slower. It's the easiest, but not really the most elegant, precise, or satisfying solution.

Also just noting that I noticed two new ships in human space that I need to adjust. The Sunder and the Hogshead.

Quantumshark commented:
> I've been thinking about this, and the comments about how overall, this PR kind of makes things faster and a bit harder to control. I'm wondering if perhaps the overal thrust levels should be reduced a bit to cut all the relative speeds down somewhat. I'm thinking maybe 10% or thereabouts. I'm going to spend some time actually playing it before I do any changes though.

I'm not sure it's entirely the raw numbers that are the issue. I played with the Splinter with this PR, and with a custom Entasi build on Azure's Avgi PR. Both actually had quite similar mobility stats - about 1,000 top speed, 100 turning, and 200 acceleration, but I found that the Splinter felt much more awkward and difficult to control. (I thought I had talked about that here at the time, but it certainly looks like I haven't, so either I imagined it or I typed it all in and forgot to press the comment button. Sorry.)
I have two hypotheses to explain the difference:
1: The changes to drag that exist on Delta but not on Azure's Avgi PR make ships with that kind of handling stats awkward to control (but weren't much of an issue for me when flying the smaller ships, and quite possibly also not when flying ships with handling stats like those that exist on the main repository, though the AI does sometimes seem to struggle with them too).
2: Those stats are fine when fighting human ships with the mobility they have in the base game currently, but feel far less pleasant to control when fighting those same ships with their buffs here.
(Or there's also the possibility that it's 3: a combination of 1 and 2, or that it's something else.)

Zitchas commented:
Hmmm... Things to ponder, for sure.

The maneuverability changes I just pushed reduces thrust and turn to 75% of their previous values. (only human space for testing purposes). I haven't tested extensively yet, but after testing 10% reduction it felt like it needed a bit more. 25% reduction feels good on initial test, and a general survey of human space numbers they pull things back down into more generally reasonable ranges.

I must admit, I am in part liking the idea that it is better to be a bit too slow and let people play in 3x mode if they want to be faster, rather than having things be faster and leave people wanting a non-existant half speed mode. Since we have an easy and convenient way to speed things up, I'm not worried about ships being a bit too slow.

I did note a freighter or two with single digit turns, which is something I want to test. That seems a touch problematic.

Zitchas commented:
OK, the main thing this needs is testing and general being-played-with. So baring any problems turning up in the near future, I am planning on extending out the 25% reduction in thrust and turn to the rest of the game, see how things look, and probably merge so it can be tested as part of other changes and experimentation.

Zitchas commented:
Currently doing an in-game pass over ship stat ranges with an eye to just making sure each faction's thrusters look like they're in basically the right ballpark.
